samples/SortBenchmark/SortBenchmark.fs

```diff
@@ -552,36 +631,103 @@ type RemoteFunc( filePartNum:int, records:int64, _dim:int , partNumS1:int, partN
+
+                            let partitionNum = ref 0
+                            if r.[0] > 0 then
+                                
+                                let iBuf = new MemoryStreamB()
+                                iBuf.WriteByte(byte 0)
+                                iBuf.Write(oBuf,0,r.[0])
+                                partitionNum := !partitionNum + 1
+                                iBuf.Seek(0L,SeekOrigin.Begin) |> ignore
+                                yield iBuf
+                        
+                            for i = 1 to nump-1 do
+                                if r.[i] > r.[i-1] then
+                                    let toCopyLen = (r.[i]-r.[i-1])
+                                    let tl = ref 0
+
+                                    let iBuf = new MemoryStreamB(toCopyLen)
+                                    iBuf.WriteByte(byte i)
+                                    iBuf.Write(oBuf,r.[i-1],toCopyLen)
+                                    partitionNum := !partitionNum + 1
+                                    yield iBuf
+
+
+
+                            let t3 = DateTime.UtcNow
+                            Interlocked.Decrement(x.repartitionThread) |> ignore
+                            RemoteFunc.sharedMem.Enqueue(oBuf)
+                            Logger.LogF( LogLevel.MildVerbose, (fun _ -> sprintf "repartition: %d records, into %d partitions, takes %f / %f s; # repartition threads: %d; shareMem Len: %d; new allocated mem: %A" (size / 100) !partitionNum ((t2-t1).TotalSeconds) ((t3-t1).TotalSeconds) !x.repartitionThread RemoteFunc.sharedMem.Count bnewbuf)  )                  
+
+                        }
+            retseq
+        elif size = 1 then
+                Logger.LogF( LogLevel.Info, ( fun _ -> sprintf "Send Flush Buffer signal from repartition"))
+                
+                seq {
+                    let nump = 
+                        if stage = RepartitionStage.StageOne then
+                            partNumS1
+                        else 
+                            partNumS2 
+                    for i = 0 to nump-1 do
+                        let iBuf = new MemoryStreamB()
+                        iBuf.WriteByte((byte) i)
+                        iBuf.WriteByte((byte) 1)
+                        yield iBuf  
+                    }      
+        else 
+            Seq.empty
 
-    member x.Repartition (stage:int) (buffer:byte[], size:int) = 
+
+
+
+    // repartition, use per-allocated memory. For testing and comparing to MemoryStreamB only, cannot be used in remote server
+    member x.RepartitionSharedMemory (stage:RepartitionStage) (buffer:byte[], size:int) = 
         if size > 0 then
             
             let retseq = seq {
                         let partionBoundary =
-                            if stage = 1 then
+                            if stage = RepartitionStage.StageOne then
                                 stageOnePartionBoundary
```

Comment:
If the stage is never sent over the wire, there's an advantage to *not* making it an enum (by ommiting the "= 1", "= 2" in the stage definition). The advantage is that the "match" keyword will check that the alternatives are exhaustive at compile time. As this is, if a "StageThree" is ever introduced, we'll have to go hunting for where to update. If we use an F# union and "match", the compiler does that for us.